### PR TITLE
Add CUDA Primitives for Population Count

### PR DESCRIPTION
**Pull request by njwhite**
Ends up as [this](http://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__INT.html#group__CUDA__MATH__INTRINSIC__INT_1g43c9c7d2b9ebf202ff1ef5769989be46)

### CODE REVIEW

File: numba/cuda/tests/cudapy/test_intrinsics.py [hunk 1]
```diff
@@ -183,6 +187,18 @@ def foo(out):
 
         self.assertTrue(np.all(arr))
 
+    def test_popc_u4(self):
+        compiled = cuda.jit("void(int32[:], uint32)")(simple_popc)
+        ary = np.zeros(1, dtype=np.int32)
+        compiled(ary, 0xF0)
+        self.assertTrue(ary[0] == 4)
```
Comment (sklam) [inline numba/cuda/tests/cudapy/test_intrinsics.py]: can you use `self.assertEqual(ary[0], 4)` instead?

Comment (sklam) [reply numba/cuda/tests/cudapy/test_intrinsics.py]: also, a bit more testing to show that popc doesn't always return 4.

Comment (njwhite) [reply numba/cuda/tests/cudapy/test_intrinsics.py]: @sklam done!

File: numba/cuda/stubs.py [hunk 2]
```diff
@@ -243,6 +243,17 @@ def local_array(shape, dtype):
     return ir.Intrinsic(fname, sig, args=(shape, dtype))
 
 
+class popc(Stub):
```
Comment (stuartarchibald) [inline numba/cuda/stubs.py]: Perhaps this should be added somewhere below in a new section as this section is marked `local_array` https://github.com/njwhite/numba/blob/dbb8471299795681cf124e0105ba00215a348169/numba/cuda/stubs.py#L234 ?

**Comment by stuartarchibald** [comment]
Thanks for the patch. This API addition should be documented (probably here https://github.com/numba/numba/blob/master/docs/source/cuda-reference/kernel.rst) before merge.

File: numba/cuda/stubs.py (hunk 2 above)
Comment (njwhite) [reply numba/cuda/stubs.py]: Done!

**Comment by njwhite** [comment]
Thanks @stuartarchibald - I've added some docs (and some new intrinsics, which will work once numba/llvmlite#339 merges. Do you know why I had to add `@lower(stubs.clz, types.u4)` to `ptx_clz_i4`? For some reason, although the `test_clz_u4` declares the kernel as `"void(int32[:], int32)"`, when it gets called with `0x0010_0000` it still gets lowered as a `u4`?

**Comment by stuartarchibald** [comment]
Answering https://github.com/numba/numba/pull/2824#issuecomment-374413928.
I think it is because the typing definition states that `clz` has signature `types.int32(types.uint32)`,
https://github.com/numba/numba/pull/2824/files#diff-a844ed5f7aa24c577c5550d1762a8506R130
and the lowering is/was defined as  `@lower(stubs.clz, types.i4)` ?

File: numba/cuda/cudaimpl.py [hunk 3]
```diff
@@ -244,6 +244,57 @@ def ptx_threadfence_device(context, builder, sig, args):
...
+
+
+@lower(stubs.popc, types.u8)
+def ptx_popc_u8(context, builder, sig, args):
+    """
+    LLVM IR implementation as suggested by `IntrinsicsNVVM.td`
+    """
+    return builder.trunc(
+        builder.ctpop(args[0]),
+        Type.int(32))
```
Comment (stuartarchibald) [inline numba/cuda/cudaimpl.py]: I think these types need to be a Numba type, not an llvmlite one? Cannot get the code to run at present as this const can't be lowered?

**Comment by njwhite** [comment]
Fixed merge conflict

File: numba/cuda/tests/cudapy/test_intrinsics.py [hunk 4]
```diff
@@ -226,6 +242,99 @@ def foo(out):
...
+        compiled = cuda.jit("void(int32[:], int32)")(simple_clz)
+        ary = np.zeros(1, dtype=np.int32)
+        compiled(ary, 0x00100000)
+        self.assertEquals(ary[0], 11)
+
+    def test_clz_u4(self):
+        """
+        Although the CUDA Math API (http://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__INT.html)
+        only says int32 & int64 arguments are supported in C code, the LLVM
+        IR input supports i8, i16, i32 & i64 (LLVM doesn't have a concept of
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_intrinsics.py]: bracket needs closing.

**Comment by stuartarchibald** [comment]
@njwhite Please could you resolve the conflict by **merging** Numba master into this branch? Thanks.

File: numba/cuda/tests/cudapy/test_intrinsics.py (hunk 4 above)
Comment (njwhite) [reply numba/cuda/tests/cudapy/test_intrinsics.py]: Done!

**Comment by njwhite** [comment]
@stuartarchibald fixed & merged!

File: numba/cuda/cudadecl.py [hunk 5]
```diff
@@ -105,6 +105,69 @@ class Cuda_threadfence_system(ConcreteTemplate):
...
+        signature(types.uint16, types.uint16),
+        signature(types.uint32, types.uint32),
+        signature(types.uint64, types.uint64),
+    ]
+
+
+@intrinsic
+class Cuda_ffs(ConcreteTemplate):
+    """
+    Supported types from http://docs.nvidia.com/cuda/nvvm-ir-spec/index.html#bit-manipulations-intrinics
```
Comment (stuartarchibald) [inline numba/cuda/cudadecl.py]: This link doesn't appear to have a reference to `ffs`?

Also, the docstring https://github.com/numba/numba/pull/2824/files#diff-1c116a52e2457a8cebe1e9d2072d195dR320 suggests 32bit only, but there's 64bit types below?

**Comment by stuartarchibald** [comment]
I've also tagged this for the `0.38 RC`, assuming the build passes, I think this should make it.

File: docs/source/cuda-reference/kernel.rst [hunk 6]
```diff
@@ -167,7 +167,34 @@ are guaranteed to not move across the memory fences by optimization passes.
...
+   Returns the number of set bits in the given value.
+
+.. function:: numba.cuda.brev
+
+   Reverses the bit pattern of an integer value, for example 0b10110110
+   becomes 0b01101101.
+
+.. function:: numba.cuda.clz
+
+   Counts the number of leading zeros in a variable.
```
[the remark this answers is not in the document]
Comment (njwhite) [reply docs/source/cuda-reference/kernel.rst]: Fixed!

File: numba/cuda/cudaimpl.py [hunk 7]
```diff
@@ -244,6 +244,50 @@ def ptx_threadfence_device(context, builder, sig, args):
...
+def ptx_clz(context, builder, sig, args):
+    return builder.ctlz(
+        args[0],
+        context.get_constant(types.boolean, 0))
+
+
+@lower(stubs.ffs, types.Any)
+def ptx_ffs(context, builder, sig, args):
+    """
+    LLVM IR implementation as suggested by `IntrinsicsNVVM.td`
```
[the remark this answers is not in the document]
Comment (njwhite) [reply numba/cuda/cudaimpl.py]: I've removed this - while the [LLVM source](https://github.com/llvm-mirror/llvm/blob/fb4ee0d83c55e3d7ccfd4e1254f3e5341826b5f4/include/llvm/IR/IntrinsicsNVVM.td#L21-L22) does have hints for converting the old `nvvm` LLVM intrinsics into new LLM intrinsics, the standard `cltz` / `cttz` intrinsics work with a wider range of integers.

File: numba/cuda/cudadecl.py (hunk 5 above)
Comment (njwhite) [reply numba/cuda/cudadecl.py]: I've made the docs a bit clearer - nb. the CUDA intrinsics are named after the PTX intrinsic (so `ffs`), not the LLVM ones (`cttz` here). Where there's not a 1:1 mapping (like `brev` currently) we have to use the CUDA name, and using a mixture of CUDA & LLVM variables for the intrinsics wouldn't be very user-friendly!

**Comment by njwhite** [comment]
@stuartarchibald thanks for review this - I've fixed all the outstanding issues

**Comment by stuartarchibald** [comment]
Thanks @njwhite, much appreciated. I'm happy with this as is the buildfarm! Merging.

**Pull request merged**